src/memory.rs

```diff
@@ -208,14 +225,82 @@ impl InMemory {
     }
 }
 
+enum InMemoryUploadState {
+    Idle,
+    Busy(JoinHandle<Result<(), io::Error>>),
+}
+
+struct InMemoryUpload {
+    location: Path,
+    data: Vec<u8>,
+    storage: Arc<RwLock<BTreeMap<Path, Bytes>>>,
+    state: InMemoryUploadState,
+}
+
+#[async_trait]
+impl MultiPartUpload for InMemoryUpload {
+    async fn abort(&mut self) -> Result<()> {
+        Ok(())
+    }
+}
+
+impl AsyncWrite for InMemoryUpload {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> std::task::Poll<Result<usize, io::Error>> {
+        self.data.extend_from_slice(buf);
+        Poll::Ready(Ok(buf.len()))
+    }
+
+    fn poll_flush(
+        self: Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), io::Error>> {
+        Poll::Ready(Ok(()))
```

Comment:
I think we *could* do that. Though doing the write in `poll_shutdown` aligns better with the behavior of multi-part upload, since we don't commit the parts until the shutdown method there. In local filesystem, we do flush partial results to the file, so that's the one case where the old value isn't readable during the upload. So I'm inclined to keep final write / commit of data in `poll_shutdown`, but let me know if you still feed differently.